File: handy-ts/quiz/q9_valid_anagram_easy.ts
/**
 * Q9) Valid Anagram
 *
 * AI-BEST:
 * Frequency counting.
 * Time: O(n), Space: O(k)
 *
 * AI-EASY:
 * Sort and compare.
 * Time: O(n log n), Space: depends on runtime
 */

function isAnagramBest(s: string, t: string): boolean {
  if (s.length !== t.length) return false;

  const counts = new Map<string, number>();
  for (const ch of s) counts.set(ch, (counts.get(ch) ?? 0) + 1);

  for (const ch of t) {
    const next = (counts.get(ch) ?? 0) - 1;
    if (next < 0) return false;
    counts.set(ch, next);
  }

  for (const value of counts.values()) {
    if (value !== 0) return false;
  }
  return true;
}

function isAnagramEasy(s: string, t: string): boolean {
  if (s.length !== t.length) return false;
  return s.split("").sort().join("") === t.split("").sort().join("");
}

function runDemo(): void {
  console.log("Q9: Valid Anagram");
  const samples: Array<[string, string]> = [
    ["anagram", "nagaram"],
    ["rat", "car"],
    ["listen", "silent"],
  ];
  for (const [s, t] of samples) {
    console.log(`Input: s="${s}", t="${t}"`);
    console.log("  BEST:", isAnagramBest(s, t));
    console.log("  EASY:", isAnagramEasy(s, t));
  }
}

runDemo();

export {};

File: handy-java/quiz/q9_valid_anagram_easy.java
import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;

/**
 * Q9) Valid Anagram
 *
 * AI-BEST:
 * Frequency counting map.
 * Time: O(n), Space: O(k)
 *
 * AI-EASY:
 * Sort both strings and compare.
 * Time: O(n log n), Space: depends on sorting implementation
 */
public class q9_valid_anagram_easy {
    static boolean isAnagramBest(String s, String t) {
        if (s.length() != t.length()) return false;

        Map<Character, Integer> counts = new HashMap<>();
        for (char ch : s.toCharArray()) {
            counts.put(ch, counts.getOrDefault(ch, 0) + 1);
        }
        for (char ch : t.toCharArray()) {
            int next = counts.getOrDefault(ch, 0) - 1;
            if (next < 0) return false;
            counts.put(ch, next);
        }

        for (int value : counts.values()) {
            if (value != 0) return false;
        }
        return true;
    }

    static boolean isAnagramEasy(String s, String t) {
        if (s.length() != t.length()) return false;

        char[] a = s.toCharArray();
        char[] b = t.toCharArray();
        Arrays.sort(a);
        Arrays.sort(b);
        return Arrays.equals(a, b);
    }

    public static void main(String[] args) {
        System.out.println("Q9: Valid Anagram");
        String[][] samples = {
                {"anagram", "nagaram"},
                {"rat", "car"},
                {"listen", "silent"}
        };
        for (String[] sample : samples) {
            String s = sample[0];
            String t = sample[1];
            System.out.println("Input: s=\"" + s + "\", t=\"" + t + "\"");
            System.out.println("  BEST: " + isAnagramBest(s, t));
            System.out.println("  EASY: " + isAnagramEasy(s, t));
        }
    }
}


